File: app/src/main/java/net/xaethos/trackernotifier/fragments/BaseResourceFragment.java
package net.xaethos.trackernotifier.fragments;

import android.content.Context;
import android.support.v4.app.Fragment;

import net.xaethos.trackernotifier.models.Resource;

import rx.Observable;

public abstract class BaseResourceFragment<T extends Resource> extends Fragment {
    private Observable<T> mResourceObservable;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof ResourceSource) {
            mResourceObservable = ((ResourceSource) context).getResourceObservable();
        }
    }

    protected Observable<T> observeResource() {
        return mResourceObservable;
    }

    public interface ResourceSource<R extends Resource> {
        Observable<R> getResourceObservable();
    }
}

File: app/src/main/java/net/xaethos/trackernotifier/fragments/BaseResourceFragment.kt
package net.xaethos.trackernotifier.fragments

import android.content.Context
import android.support.v4.app.Fragment

import net.xaethos.trackernotifier.models.Resource

import rx.Observable

abstract class BaseResourceFragment<T : Resource> : Fragment() {
    protected var resourceObservable: Observable<T>? = null
        private set

    override fun onAttach(context: Context) {
        super.onAttach(context)
        resourceObservable = (context as? ResourceSource<T>)?.resourceObservable
    }

    interface ResourceSource<R : Resource> {
        val resourceObservable: Observable<R>
    }
}
